File: community.variability/R/community.variability/R/community_metrics.R
## ================================================================
## Community array construction and metric output helpers
## ================================================================
##
## Data cleaning can remain tidy/data-frame oriented. Once the
## observation table is prepared, these helpers convert it to the
## numerical metacommunity array used by the variability functions:
##
##   X[time, site, taxon]
##
## where:
##   time  = repeated surveys t
##   site  = local communities i
##   taxon = species/taxa j
##
## Metric output order follows the alpha-gamma-phi partitioning:
##   gamma = regional/metacommunity variability
##   alpha = local/community variability
##   phi   = spatial synchrony component
##
## BD_spatial_weighted is reported as BD_beta here only if you want
## to keep the previous output name. It is a biomass-weighted summary
## of spatial compositional differences among sites through time, not
## the Lamy et al. BD_phi ratio.
community_metric_order <- c(
  "CV_gamma", "CV_alpha", "CV_phi",
  "BD_gamma", "BD_alpha", "BD_phi", "BD_beta"
)
make_community_array <- function(data_wide,
                                 taxa_cols,
                                 time_step_col_name = "timestep",
                                 site_id_col_name = "site") {
  ## Identify the time and site levels that define the first two
  ## dimensions of X. Sorting gives reproducible array ordering.
  time_ids <- sort(unique(data_wide[[time_step_col_name]]))
  site_ids <- sort(unique(data_wide[[site_id_col_name]]))
  ## Initialize the biomass array:
  ##   X[t, i, j] = biomass of taxon j at site i and time t
  X <- array(
    0,
    dim = c(length(time_ids), length(site_ids), length(taxa_cols)),
    dimnames = list(
      timestep = as.character(time_ids),
      site = as.character(site_ids),
      taxon = taxa_cols
    )
  )
  ## Map each row in the prepared wide table to its array position.
  time_index <- match(data_wide[[time_step_col_name]], time_ids)
  site_index <- match(data_wide[[site_id_col_name]], site_ids)
  ## Extract taxon biomass columns as a numeric matrix.
  ## Non-finite values are treated as zero biomass.
  biomass <- as.matrix(data_wide[, taxa_cols, drop = FALSE])
  storage.mode(biomass) <- "double"
  biomass[!is.finite(biomass)] <- 0
  ## Fill X row by row.
  ##
  ## If duplicate site x time rows occur, their biomass values are
  ## added. This preserves total biomass and avoids silently dropping
  ## replicate records.
  for (row_idx in seq_len(nrow(data_wide))) {
    X[time_index[row_idx], site_index[row_idx], ] <-
      X[time_index[row_idx], site_index[row_idx], ] + biomass[row_idx, ]
  }
  X
}
calc_metacommunity_metrics <- function(X) {
  ## Calculate alpha and gamma components first.
  ##
  ## Aggregate variability:
  ##   CV_gamma = regional variability in total metacommunity biomass
  ##   CV_alpha = local variability in total site biomass
  ##
  ## Compositional variability:
  ##   BD_gamma = regional variability in metacommunity composition
  ##   BD_alpha = biomass-weighted local compositional variability
  metrics <- c(
    CV_gamma = CV_gamma(X),
    CV_alpha = CV_alpha(X),
    BD_gamma = BD_gamma(X),
    BD_alpha = BD_alpha(X),
    BD_beta = BD_spatial_weighted(X)
  )
  ## Calculate spatial synchrony components using the multiplicative
  ## partitions:
  ##
  ##   CV_gamma = CV_alpha * CV_phi
  ##   BD_gamma = BD_alpha * BD_phi
  metrics <- c(
    metrics,
    CV_phi = metrics[["CV_gamma"]] / metrics[["CV_alpha"]],
    BD_phi = metrics[["BD_gamma"]] / metrics[["BD_alpha"]]
  )
  ## Return a tidy metric table in a fixed order for downstream joins,
  ## plotting, and comparison across datasets.
  tibble::tibble(
    varname = community_metric_order,
    estimate = unname(metrics[community_metric_order])
  )
}
wide_metric_table <- function(metric_table,
                              value_col = "estimate") {
  ## Convert the long metric table:
  ##   varname | estimate
  ##
  ## to a one-row wide table:
  ##   CV_gamma | CV_alpha | CV_phi | ...
  metric_table %>%
    dplyr::select(varname, dplyr::all_of(value_col)) %>%
    tidyr::pivot_wider(
      names_from = varname,
      values_from = dplyr::all_of(value_col)
    )
}
calc_spatial_bd_by_time <- function(X) {
  ## Return the time-resolved spatial compositional variability
  ## calculated inside spatial_bd_by_time().
  ##
  ## BD:
  ##   spatial compositional variability among sites at each timestep
  ##
  ## total_metacommunity_biomass:
  ##   X_.t., used to weight each timestep
  ##
  ## weights:
  ##   w_t = X_.t. / sum_t X_.t.
  ##
  ## BD_x_wt:
  ##   weighted contribution of each timestep to BD_spatial_weighted
  site_biomass_by_time <- apply(X, c(1, 2), sum, na.rm = TRUE)
  site_relative_biomass <- sweep(
    X,
    c(1, 2),
    site_biomass_by_time,
    "/"
  )
  site_relative_biomass[!is.finite(site_relative_biomass)] <- 0
  site_hellinger <- sqrt(site_relative_biomass)
  BD <- vapply(seq_len(dim(site_hellinger)[1]), function(time_idx) {
    site_by_taxon <- site_hellinger[time_idx, , , drop = FALSE][1, , ]
    taxon_spatial_var <- apply(site_by_taxon, 2, stats::var, na.rm = TRUE)
    sum(taxon_spatial_var, na.rm = TRUE)
  }, numeric(1))
  total_metacommunity_biomass <- rowSums(site_biomass_by_time, na.rm = TRUE)
  weights <- total_metacommunity_biomass /
    sum(total_metacommunity_biomass, na.rm = TRUE)
  tibble::tibble(
    timestep = dimnames(X)$timestep,
    BD = BD,
    total_metacommunity_biomass = total_metacommunity_biomass,
    weights = weights,
    BD_x_wt = BD * weights
  )
}

File: community.variability/python/community_variability/community_metrics.py
"""Community array construction and metric output helpers."""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass

import numpy as np
import pandas as pd

from .variability import bd_alpha, bd_gamma, bd_spatial_weighted, cv_alpha, cv_gamma


COMMUNITY_METRIC_ORDER = [
    "CV_gamma",
    "CV_alpha",
    "CV_phi",
    "BD_gamma",
    "BD_alpha",
    "BD_phi",
    "BD_beta",
]


@dataclass(frozen=True)
class CommunityArray:
    """Numerical community array with labels for X[time, site, taxon]."""

    values: np.ndarray
    timestep: list[str]
    site: list[str]
    taxon: list[str]

    @property
    def dimnames(self) -> dict[str, list[str]]:
        return {"timestep": self.timestep, "site": self.site, "taxon": self.taxon}


def make_community_array(
    data_wide: pd.DataFrame,
    taxa_cols: Sequence[str],
    time_step_col_name: str = "timestep",
    site_id_col_name: str = "site",
) -> CommunityArray:
    """Build X[t, i, j] from one wide biomass row per time x site sample."""
    time_ids = sorted(data_wide[time_step_col_name].dropna().unique().tolist())
    site_ids = sorted(data_wide[site_id_col_name].dropna().unique().tolist())
    taxa_cols = list(taxa_cols)

    # Initialize the biomass array:
    # X[t, i, j] = biomass of taxon j at site i and time t.
    X = np.zeros((len(time_ids), len(site_ids), len(taxa_cols)), dtype=float)
    time_index = {value: idx for idx, value in enumerate(time_ids)}
    site_index = {value: idx for idx, value in enumerate(site_ids)}

    # Extract taxon biomass columns as a numeric matrix.
    # Non-finite values are treated as zero biomass.
    biomass = data_wide.loc[:, taxa_cols].apply(pd.to_numeric, errors="coerce").to_numpy(dtype=float).copy()
    biomass[~np.isfinite(biomass)] = 0.0

    # Fill X row by row. Duplicate site x time rows are added so total
    # biomass is preserved rather than silently dropping replicate records.
    for row_idx, row in data_wide.reset_index(drop=True).iterrows():
        t = time_index[row[time_step_col_name]]
        i = site_index[row[site_id_col_name]]
        X[t, i, :] += biomass[row_idx, :]

    return CommunityArray(
        values=X,
        timestep=[str(value) for value in time_ids],
        site=[str(value) for value in site_ids],
        taxon=[str(value) for value in taxa_cols],
    )


def calc_metacommunity_metrics(X) -> pd.DataFrame:
    """Calculate alpha-gamma-phi aggregate and compositional partitions."""
    # Calculate alpha and gamma components first.
    # CV_gamma and CV_alpha use aggregate biomass; BD_gamma and BD_alpha use
    # Hellinger composition, z = sqrt(relative biomass).
    metrics = {
        "CV_gamma": cv_gamma(X),
        "CV_alpha": cv_alpha(X),
        "BD_gamma": bd_gamma(X),
        "BD_alpha": bd_alpha(X),
        "BD_beta": bd_spatial_weighted(X),
    }

    # Multiplicative partitions:
    # CV_gamma = CV_alpha * CV_phi
    # BD_gamma = BD_alpha * BD_phi
    with np.errstate(divide="ignore", invalid="ignore"):
        metrics["CV_phi"] = float(np.divide(metrics["CV_gamma"], metrics["CV_alpha"]))
        metrics["BD_phi"] = float(np.divide(metrics["BD_gamma"], metrics["BD_alpha"]))

    return pd.DataFrame(
        {
            "varname": COMMUNITY_METRIC_ORDER,
            "estimate": [metrics[name] for name in COMMUNITY_METRIC_ORDER],
        }
    )


def wide_metric_table(metric_table: pd.DataFrame, value_col: str = "estimate") -> pd.DataFrame:
    """Convert long metric output to one wide row."""
    return metric_table.pivot_table(columns="varname", values=value_col, aggfunc="first").reset_index(drop=True)


def calc_spatial_bd_by_time(X) -> pd.DataFrame:
    """Return time-resolved spatial compositional variability."""
    values = X.values if hasattr(X, "values") else X
    values = np.asarray(values, dtype=float)
    if values.ndim != 3:
        raise ValueError("Community data must have shape X[time, site, taxon].")

    site_biomass_by_time = np.nansum(values, axis=2)
    with np.errstate(divide="ignore", invalid="ignore"):
        site_relative_biomass = values / site_biomass_by_time[:, :, None]
    site_relative_biomass = np.where(np.isfinite(site_relative_biomass), site_relative_biomass, 0.0)
    site_hellinger = np.sqrt(site_relative_biomass)

    # Spatial variability per timestep:
    # BD_t^h = sum_j Var_i(z_itj).
    taxon_spatial_var = np.nanvar(site_hellinger, axis=1, ddof=1)
    bd = np.nansum(taxon_spatial_var, axis=1)

    total_metacommunity_biomass = np.nansum(site_biomass_by_time, axis=1)
    weights = total_metacommunity_biomass / np.nansum(total_metacommunity_biomass)
    timestep = X.timestep if hasattr(X, "timestep") else [str(i) for i in range(values.shape[0])]

    return pd.DataFrame(
        {
            "timestep": timestep,
            "BD": bd,
            "total_metacommunity_biomass": total_metacommunity_biomass,
            "weights": weights,
            "BD_x_wt": bd * weights,
        }
    )
